identity/migration/src/main/java/io/camunda/identity/migration/AuthorizationMigrationHandler.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Camunda License 1.0. You may not use this file
+ * except in compliance with the Camunda License 1.0.
+ */
+package io.camunda.identity.migration;
+
+import io.camunda.identity.migration.dto.UserResourceAuthorization;
+import io.camunda.service.AuthorizationServices;
+import io.camunda.service.AuthorizationServices.PatchAuthorizationRequest;
+import io.camunda.zeebe.protocol.impl.record.value.authorization.AuthorizationRecord;
+import io.camunda.zeebe.protocol.record.value.AuthorizationResourceType;
+import io.camunda.zeebe.protocol.record.value.PermissionAction;
+import io.camunda.zeebe.protocol.record.value.PermissionType;
+import java.util.AbstractMap.SimpleEntry;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.stream.Collectors;
+import org.springframework.stereotype.Component;
+
+@Component
+public class AuthorizationMigrationHandler {
+
+  private final AuthorizationServices<AuthorizationRecord> authorizationService;
+  private final ManagementIdentityProxy managementIdentityProxy;
+
+  public AuthorizationMigrationHandler(
+      final AuthorizationServices<AuthorizationRecord> authorizationService,
+      final ManagementIdentityProxy managementIdentityProxy) {
+    this.authorizationService = authorizationService;
+    this.managementIdentityProxy = managementIdentityProxy;
+  }
+
+  public void migrate() {
+    UserResourceAuthorization lastRecord = null;
+    while (true) {
+      final List<UserResourceAuthorization> authorizations =
+          managementIdentityProxy.fetchUserResourceAuthorizations(lastRecord, 100);
+      if (authorizations.isEmpty()) {
+        return;
+      }
+      lastRecord = authorizations.getLast();
+      final Map<String, Map<String, Map<String, List<String>>>> authorizationMap =
```

Comment:
❓ I'm concerned at the complexity here (and in the following `forEach`) but I think that we are somewhat limited in our options so it may be that we actually have no other way.

I wonder if we can build a composite key (represented by a internal record) that maps to the list of resource IDs, so instead of 
```
{"ben", {"process-definition", {"write", ["resource-one", "resource-two"]}}}
```
we could have 
```
{<composite key object>, ["resource-one", "resource-two"]}
```

Composite key represented by the class 
```
  private record AuthorizationCompositeKey(
      String username, String resourceType, String permission) {}
```

I think that would drop the stream to a shorter size because we would say
```
      final Map<AuthorizationCompositeKey, List<String>> authorizationsMap =
          authorizations.stream()
              .collect(
                  Collectors.groupingBy(
                      auth ->
                          new AuthorizationCompositeKey(
                              auth.username(), auth.resourceType(), auth.permission()),
                      Collectors.mapping(
                          UserResourceAuthorization::resourceId, Collectors.toList())));
```

This also means the stream below with the conversion simplifies because we have the composite key giving us data and the linked resources i.e. 
```
      authorizationsMap.forEach(
          (key, value) -> {
            final long ownerKey = getOwnerKeyForUsername(key.username());
            authorizationService.patchAuthorization(
                new PatchAuthorizationRequest(
                    ownerKey,
                    PermissionAction.ADD,
                    convertResourceType(key.resourceType()),
                    //helper method to expand key.permission after conversion));
          });
```

Reply:
I like the idea to use the CompositKey but for our case the composite key should n't contains permissions as we are patching all permissions + (resourceIds) which belong to a owner/resource in one call, so we can just decrease one level of the map and put owner + resource Type together as a key.